locallib.php

```diff
@@ -475,5 +477,45 @@ function auth_saml2_get_default_idp() {
     return $defaultidp;
 }
 
+/**
+ * This helper function processes the regenerate form.
+ * Moved here so we can use it in a unit test.
+ *
+ * @param $fromform
+ * @return void|string
+ */
+function auth_saml2_process_regenerate_form($fromform) {
+    global $CFG, $USER;
+    $dn = array(
+        'commonName' => substr($fromform->commonname, 0, 64),
+        'countryName' => $fromform->countryname,
+        'emailAddress' => $fromform->email,
+        'localityName' => $fromform->localityname,
+        'organizationName' => $fromform->organizationname,
+        'stateOrProvinceName' => $fromform->stateorprovincename,
+        'organizationalUnitName' => $fromform->organizationalunitname,
+    );
+    $numberofdays = $fromform->expirydays;
+
+    $saml2auth = new \auth_plugin_saml2();
+    $eventdata = [
+        'reason' => "regenerated in saml settings page",
+        'userid' => $USER->id,
+    ];
+    cert_regenerated::create(['other' => $eventdata])->trigger();
```

Comment:
should we swap this order and only emit the event if the cert creation was a success?

Reply:
I suppose since an error would be logged anyway, we don't need the event to be emitted as well in that case. I have made that change now